### PR TITLE
Tt 138 create time sync library

### PR DESCRIPTION
**Pull request by paulbourelly999**
# PR Details
## Description
This is a new C++ library intended to replace the mmitss [TimeStamp](https://github.com/usdot-fhwa-stol/mmitss/blob/develop/mmitss/src/common/Timestamp.cpp) functions with a class that will allow mmitss services to sync to simulation time. Below is a diagram highlighting how this is currently planned
image

## Related Jira Key
[TT-138](https://usdot-carma.atlassian.net/browse/TT-138)

## Motivation and Context
Allow for MMITSS Services Time Sync

## How Has This Been Tested?
Unit testing

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

[TT-138]: https://usdot-carma.atlassian.net/browse/TT-138?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/workflows/sonar-scanner.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
+on:
+  workflow_dispatch:
+  # Trigger analysis when pushing in master or pull requests, and when creating
+  # a pull request.
+  push:
+   branches:
+     - 'master'
+     - 'develop'
+     - 'release/*'
```
Comment (dan-du-car) [inline .github/workflows/sonar-scanner.yml:9]: Do we need a feature branch?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+on:
+  workflow_dispatch:
+  push:
+   branches:
+     - master
+     - develop
+     - 'release/*'
+     - 'feature*'
+     - 'tt-138-create-time-sync-library'
```
Comment (dan-du-car) [inline .github/workflows/build.yml]: Is this going to be remove  prior to merge?

File: .vscode/launch.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": "gdb",
```
Comment (dan-du-car) [inline .vscode/launch.json]: Maybe include `Test` in the name to be clear?

File: .vscode/tasks.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{
+    // See https://go.microsoft.com/fwlink/?LinkId=733558
+    // for the documentation about the tasks.json format
+    "version": "2.0.0",
+    "tasks": [
+        {
+            "label": "build",
+            "type": "shell",
+            "command": "./build.sh -d",
```
Comment (dan-du-car) [inline .vscode/tasks.json:9]: Is this -d debug? Could the label be a bit clearer?

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+)
+# RPATH required for allowing dpkg-shlibdeps to find stol-apt-repo dependencies
+set_target_properties( ${PROJECT_NAME} PROPERTIES
+    SOVERSION 1
+    INSTALL_RPATH_USE_LINK_PATH TRUE
+    INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib"
+    
+)
+
+include(CommonSource)
```
Comment (dan-du-car) [inline CMakeLists.txt:54]: Does this mean they all depends on carma-build? Is there a step to run carma-build first before it comes to build this lib?

Comment (dan-du-car) [reply CMakeLists.txt:54]: Is it possible to add inline comments to point out to users where these modules are located?

File: Doxyfile [hunk 6]
```diff
@@ -0,0 +1,2 @@
+INPUT = include/ src/
+OUTPUT_DIRECTORY       = ./docs/
```
Comment (dan-du-car) [inline Doxyfile:2]: Is this executed? Where are the generated file stored, GitHub or?

File: build.sh [hunk 7]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+set -e
+
+${CARMA_OPT_DIR}/scripts/build_script.sh -p $@ -DBUILD_TESTS=ON -DCREATE_DEB_PACKAGE=ON
```
Comment (dan-du-car) [inline build.sh:5]: Assume this build_script.sh is from carma-build repo, where in the process we download the carma-build and install it in the ${CARMA_OPT_DIR} directory?

File: include/TimeSync.hpp [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+#include <spdlog/spdlog.h>
+
+#include <carma-clock/carma_clock.h>
+#include <rapidjson/document.h>
+#include <udp-socket/UdpServer.hpp>
+#include <atomic>
+
+namespace carma_streets_time_sync {
+    /**
+     * TimeSyncMessage struct
```
Comment (dan-du-car) [inline include/TimeSync.hpp:30]: It would be more helpful to include comment about the purpose of each fields in the struct?

File: include/TimeSync.hpp [hunk 9]
```diff
@@ -0,0 +1,70 @@
+#pragma once
```
Comment (dan-du-car) [inline include/TimeSync.hpp:16]: Please add Leidos header?

File: test/python_timesync_test.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+import time
+
+
+"""Test Case for testing basic udp-time-sync functionality in python
+"""
+class TestCarmaClock(unittest.TestCase):
+
+    def testSimClockInitializeException(self):
+        timeSync = libTimeSync.TimeSync()
+        import socket
```
Comment (dan-du-car) [inline test/python_timesync_test.py]: Can we move this import to  the top?

File: test/python_timesync_test.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/python3
+import unittest
+import importlib
+libTimeSync = importlib.import_module("libudp_time_sync")
+# import libcarma_clock
```
Comment (dan-du-car) [inline test/python_timesync_test.py]: Remove the commented out code?

File: src/TimeSync.cpp [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        else {
+            SPDLOG_WARN("System config {0} was not set! Using default value {1}!" ,config_name, default_val);
+            return default_val;
+        }
+    }
+
+    TimeSyncMessage readTimeSyncMessage(const std::string &time_sync) {
+        TimeSyncMessage message;
+        rapidjson::Document document;
+        document.Parse(time_sync.c_str());
```
Comment (dan-du-car) [inline src/TimeSync.cpp:108]: Do we need a try catch in case of any parse error?

File: src/TimeSync.cpp [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        SPDLOG_INFO("Parsing time sync message: {0}", time_sync);
+        if (document.HasParseError()) {
+            throw std::runtime_error( "Parsing error: " + document.GetParseError() );
+        }
+        if (!document.IsObject() || !document.HasMember("timestamp") || !document["timestamp"].IsUint64()) {
+            throw std::runtime_error( "Parsing error: time sync message " + time_sync + " is not a valid time sync message!" );    
+        }
+        message.timestamp =  document["timestamp"].GetUint64();
+        if (!document.IsObject() || !document.HasMember("timestep") || !document["timestep"].IsUint64()) {
+            throw std::runtime_error( "Parsing error: time sync message " + time_sync + " is not a valid time sync message!" );
```
Comment (dan-du-car) [inline src/TimeSync.cpp:118]: Maybe use custom extended error class?

File: src/TimeSync.cpp [hunk 14]
```diff
@@ -0,0 +1,91 @@
+#include "TimeSync.hpp"
+
+namespace carma_streets_time_sync {
+    TimeSync::TimeSync(const std::string &ip , unsigned int port ) : initialized(false), _ip(ip), _port(port)  {}
+ 
+    void TimeSync::start() {
+        if (!initialized) {
+            // Intialize carma clock
+            std::string sim_mode_string = getSystemConfig("SIMULATION_MODE","TRUE");
```
Comment (dan-du-car) [inline src/TimeSync.cpp]: Can we add a note in the readme.md or include a sample.env file to indicator some environment variable dependencies needed for this lib?

File: .vscode/tasks.json [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+            "group": {
+                "kind": "build",
+                "isDefault": true
+            },
+            "problemMatcher": [
+                "$gcc"
+            ]
+        },
```
Comment (dan-du-car) [inline .vscode/tasks.json:17]: Is it possible to add a run task? Maybe another testing script to standup a udp server to interact with this lib if it could be integration tested independently?

**Comment by dan-du-car** [comment]
The diagram is great, but it is confusing without detailed description along with it. Is there a link to the doc for this diagram to provide more context?

File: src/TimeSync.cpp [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+            if ( _simulation_mode) {
+                // Initalize kafka consumer
+                SPDLOG_INFO("TimeSync is in simulation mode!");
+                _time_consumer= std::make_unique<udp_socket::UdpServer>(_ip, _port);
+                consumer_thread = std::thread([this]() {
+                    
+                    consumeTimeLoop();                        
+                    
+                });
+                consumer_thread.detach();
```
Comment (dan-du-car) [inline src/TimeSync.cpp:41]: Cannot we use join here?

File: .github/workflows/build.yml (hunk 2 above)
Comment (paulbourelly999) [reply .github/workflows/build.yml]: Let me remove this.

File: .github/workflows/sonar-scanner.yml (hunk 1 above)
Comment (paulbourelly999) [reply .github/workflows/sonar-scanner.yml:9]: I don't think so

File: .vscode/launch.json (hunk 3 above)
Comment (paulbourelly999) [reply .vscode/launch.json]: Sure

File: .vscode/tasks.json (hunk 4 above)
Comment (paulbourelly999) [reply .vscode/tasks.json:9]: This is a regular flag that just indicates to the CMAKE to set debug flags for line coverage etc.

File: .vscode/tasks.json (hunk 15 above)
Comment (paulbourelly999) [reply .vscode/tasks.json:17]: There is no executable. Just a library

File: Doxyfile (hunk 6 above)
Comment (paulbourelly999) [reply Doxyfile:2]: No it is not executed yet. Just a file we can use if we chose to

File: src/TimeSync.cpp (hunk 12 above)
Comment (paulbourelly999) [reply src/TimeSync.cpp:108]: Parsing does not throw and error, it sets HasParseError() (see line 93)

File: .github/workflows/sonar-scanner.yml (hunk 1 above)
Comment (paulbourelly999) [reply .github/workflows/sonar-scanner.yml:9]: added it anyway just incase

**Review by dan-du-car — APPROVED** [review]
